### PR TITLE
GetTopCounties testing

### PR DESCRIPTION
**Pull request by ice-laius**

### CODE REVIEW

File: users/top_countries_test.go [hunk 1]
```diff
@@ -38,3 +51,300 @@ func assertUsersPerCountry(ctx context.Context, tb testing.TB, country devicemet
...
+	if testing.Short() {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), testDeadline)
+	defer cancel()
+
+	const userCount = 10
+	const country = "US"
+
+	var res []*CountryStatistics
```
Comment (ice-myles) [inline users/top_countries_test.go:115]: Move `res` and `err` variables closer to the place where it is used. The same for `userCount` and `country`.

File: users/top_countries_test.go [hunk 2]
```diff
@@ -38,3 +51,300 @@ func assertUsersPerCountry(ctx context.Context, tb testing.TB, country devicemet
...
+
+func TestRepository_GetTopCountries_Success_DifferentCountries(t *testing.T) { //nolint:paralleltest // We need a clean database.
+	if testing.Short() {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), testDeadline)
+	defer cancel()
+
+	const userCount = 11
+	var res []*CountryStatistics
```
Comment (ice-myles) [inline users/top_countries_test.go:148]: The same here: we define variables in the tests closer to the places where they used.
This recommendation for all tests.

**Review by ice-myles — CHANGES_REQUESTED** [review]

File: users/top_countries_test.go [hunk 3]
```diff
@@ -23,10 +26,20 @@ func (c *CountryStatistics) bindExisting(ctx context.Context, tb testing.TB, cou
...
+	require.NoError(tb, dbConnector.PrepareExecuteTyped("SELECT country, user_count FROM users_per_country", map[string]interface{}{}, res))
+
+	return *res
+}
+
 func assertUsersPerCountry(ctx context.Context, tb testing.TB, country devicemetadata.Country, expected uint64) {
 	tb.Helper()
 	require.NoError(tb, ctx.Err())
-	windowedCtx, windowedCancel := context.WithTimeout(ctx, 2*time.Second)
+	windowedCtx, windowedCancel := context.WithTimeout(ctx, time.Duration(expected)*time.Second)
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply users/top_countries_test.go:42]: why fixed length is better then number-dependent?
I have 66 users in sorting test

File: cmd/eskimo/top_countries_test.go [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+	tb.Helper()
+	countryStats := new([]*users.CountryStatistics)
+	err := json.Unmarshal([]byte(body), countryStats)
+	require.NoError(tb, err)
+
+	return *countryStats
+}
+
+func createUsersFromDifferentCountries(ctx context.Context, tb testing.TB, userIDs, token []string) {
+	countryList := map[string]struct{}{"AL": {}, "RU": {}, "RO": {}, "US": {}, "AT": {}, "HU": {}, "GB": {}, "UA": {}, "FR": {}, "NO": {}}
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply cmd/eskimo/top_countries_test.go:166]: cuz I use the same type for creating and searching
todo move map to single place

File: users/top_countries_test.go [hunk 5]
```diff
@@ -23,10 +26,20 @@ func (c *CountryStatistics) bindExisting(ctx context.Context, tb testing.TB, cou
 	return c
 }
 
+func (c *CountryStatistics) bindExistingAll(ctx context.Context, tb testing.TB) []*CountryStatistics {
+	tb.Helper()
+	require.NoError(tb, ctx.Err())
+
+	res := new([]*CountryStatistics)
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply users/top_countries_test.go:33]: done

File: users/top_countries_test.go [hunk 6]
```diff
@@ -38,3 +51,300 @@ func assertUsersPerCountry(ctx context.Context, tb testing.TB, country devicemet
 
 	assert.Fail(tb, "unexpected users per country", "expected %v, actual %v", fmt.Sprint(expected), fmt.Sprint(actual))
 }
+
+func assertUsersCount(ctx context.Context, tb testing.TB, expected uint64) {
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply users/top_countries_test.go:55]: assertUsersPerCountry asserts for users with specified country
so should be something like assertUsersPerAllCountries

File: users/top_countries_test.go [hunk 7]
```diff
@@ -38,3 +51,300 @@ func assertUsersPerCountry(ctx context.Context, tb testing.TB, country devicemet
...
+
+	SETUP("we cleanup everything in the database", func() { mustDeleteAllUsers(ctx, t) })
+	GIVEN("create sequentially referring users", func() {
+		createNUsersWithSingleCountry(ctx, t, userCount, country)
+	})
+	GIVEN("we have some users from the default country", func() {
+		assertUsersPerCountry(ctx, t, country, userCount)
+	})
+	for _, cntry := range []string{"US", "us"} {
+		WHEN("get top countries", func() {
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply users/top_countries_test.go:126]: done

File: cmd/eskimo/top_countries_test.go [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+	"time"
+)
+
+//nolint:paralleltest // We can't parallelize it because we have a limit number of real auth tokens.
+//nolint:funlen // long is ok
+func TestService_GetTopCountries_Success_US_TenUsers(t *testing.T) {
+	if testing.Short() {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply cmd/eskimo/top_countries_test.go:23]: done

File: cmd/eskimo/top_countries_test.go [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+	defer cancel()
+
+	userIDs, token := bridge.GetAllTestingAuthorizations()
+
+	var body fixture.ActualRespBody
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply cmd/eskimo/top_countries_test.go:28]: done

File: cmd/eskimo/top_countries_test.go [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+	})
+	WHEN("get top countries for all countries", func() {
+		body, status = bridge.GetTopCountries(ctx, t, token[0], "", 1000, 0)
+	})
+	THEN(func() {
+		IT("is successfully returned", func() {
+			assert.Equal(t, 200, status)
+		})
+		IT("check body", func() {
+			countryStats := verifyGetTopCountriesBody(t, body)
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply cmd/eskimo/top_countries_test.go:45]: it should be discussed
I dont like comparing data as strings

File: cmd/eskimo/top_countries_test.go [hunk 11]
```diff
@@ -0,0 +1,224 @@
...
+}
+
+func TestService_GetTopCountries_Success_DifferentCountries(t *testing.T) {
+	if testing.Short() {
+		return
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), 60*time.Second)
+	defer cancel()
+
+	userIDs, token := bridge.GetAllTestingAuthorizations()
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply cmd/eskimo/top_countries_test.go:59]: done

File: cmd/eskimo/top_countries_test.go [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+			IT("check offset", func() {
+				require.Equal(t, userCount-offset, len(countryStats))
+			})
+		})
+	}
+}
+
+func verifyGetTopCountriesBody(tb testing.TB, body fixture.ActualRespBody) []*users.CountryStatistics {
+	tb.Helper()
+	countryStats := new([]*users.CountryStatistics)
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply cmd/eskimo/top_countries_test.go:158]: done

File: users/top_countries_test.go [hunk 13]
```diff
@@ -38,3 +51,300 @@ func assertUsersPerCountry(ctx context.Context, tb testing.TB, country devicemet
...
+	countryList := map[string]uint64{"AL": 1, "RU": 2, "RO": 3, "US": 4, "AT": 5, "HU": 6, "GB": 7, "UA": 8, "FR": 9, "NO": 10, "DE": 11}
+
+	for key, value := range countryList {
+		for j := 0; j < int(value); j++ {
+			createUserFromCountry(ctx, tb, key)
+		}
+	}
+}
+
+func verifyUsersCountry(tb testing.TB, users []*CountryStatistics, country devicemetadata.Country) {
```
[the remark this answers is not in the document]
Comment (ice-laius) [reply users/top_countries_test.go:302]: assertUsersPerCountry uses database

File: users/top_countries_test.go (hunk 1 above)
Comment (ice-laius) [reply users/top_countries_test.go:115]: done

File: users/top_countries_test.go (hunk 2 above)
Comment (ice-laius) [reply users/top_countries_test.go:148]: done

**Comment by ice-myles** [comment]
Was rebased into the `feature/fixing-tests` branch.

**Pull request closed without merging**